libraries/Microsoft.Bot.StreamingExtensions/ConcurrentStream.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.StreamingExtensions.Payloads;
+
+namespace Microsoft.Bot.StreamingExtensions
+{
+    internal class ConcurrentStream : Stream
+    {
+        private readonly ContentStreamAssembler _assembler;
+        private readonly Queue<byte[]> _bufferQueue = new Queue<byte[]>();
+
+        private readonly SemaphoreSlim dataAvailable = new SemaphoreSlim(0, int.MaxValue);
+        private readonly object syncLock = new object();
+
+        private long _producerLength = 0;       // total length
+        private long _consumerPosition = 0;     // read position
+
+        private byte[] _active = null;
+        private int _activeOffset = 0;
+
+        private bool _end = false;
+
+        internal ConcurrentStream(ContentStreamAssembler assembler)
+        {
+            _assembler = assembler;
+        }
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => true;
+
+        public override long Length => _producerLength;
+
+        public override long Position { get => _consumerPosition; set => throw new NotSupportedException(); }
+
+        public override void Flush()
+        {
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (_end)
+            {
+                return 0;
+            }
+
+            if (_active == null)
+            {
```

Comment:
This is buggy. Your locking primitives don't work with how you're checking this.